File: java/patterns/src/main/java/org/prototypepatterns/behavioral/memento/MementoPattern.java
package org.prototypepatterns.behavioral.memento;

public final class MementoPattern {
    private MementoPattern() {
    }

    public record Memento(String text) {
    }

    public static final class TextEditor {
        private String text = "";

        public void type(String value) {
            text = value;
        }

        public Memento save() {
            return new Memento(text);
        }

        public void restore(Memento memento) {
            text = memento.text();
        }

        public String text() {
            return text;
        }
    }

    public static String demo() {
        TextEditor editor = new TextEditor();
        editor.type("draft");
        Memento checkpoint = editor.save();
        editor.type("final");
        editor.restore(checkpoint);
        return editor.text();
    }
}

File: csharp/patterns/src/PrototypePatterns/Behavioral/Memento/MementoPattern.cs
namespace PrototypePatterns.Behavioral.Memento;

public sealed record Memento(string Text);

public sealed class TextEditor
{
    private string _text = string.Empty;

    public void Type(string value) => _text = value;

    public Memento Save() => new(_text);

    public void Restore(Memento memento) => _text = memento.Text;

    public string Text => _text;
}

public static class MementoPattern
{
    public static string Demo()
    {
        var editor = new TextEditor();
        editor.Type("draft");
        var checkpoint = editor.Save();
        editor.Type("final");
        editor.Restore(checkpoint);
        return editor.Text;
    }
}
